experiment: explosions subsystem

dageavtobusnick opened the pull request:
<!-- Пишите **НИЖЕ** заголовков и **ВЫШЕ** комментариев, иначе ваш текст может не отобразиться. -->
<!-- В Contributing.MD вы можете найти некоторые рекомендации к оформлению пулл-реквеста. -->

## Описание
Переносит код обработки взрыва в subsystem, по примеру https://github.com/tgstation/tgstation (не порт).
Так же сделаны микро оптимизации имеющегося кода обработки взрыва.
Взрывы обрабатываются согласно приоритету. Приоритетом является количество обрабатываемых турфов. Чем меньше взрыв, тем приоритетнее он обработается(реализовано на очереди с приоритетами на основе бинарной кучи)
Обработка турфов в рамках одного происходит последовательно(на основе очереди на связных списках). 
Так же с https://github.com/tgstation/tgstation был перенесен более оптимальный прок для отбора обрабатываемых турфов.
<!-- Опишите, что делает ваш Pull request. Документировать каждую деталь не требуется, просто укажите основные изменения. -->

## Причина создания ПР / Почему это хорошо для игры
Исполнять такую ресурсозатратную операцию как взрыв не в сабсистеме чревато проблемами. В целом перенос в сабсистему сделал обработку плавнее и стабильнее.
<!-- Здесь можно оставить ссылку на сообщение в #отчеты-по-предложениям, чтобы подтвердить, что предложение обсуждалось внутри Discord-сообщества. -->
<!-- Если отчёта нет, то укажите, почему это изменение положительно влияет на игру. -->
<!-- В случае исправления бага, укажите ссылку на канал в #баг-репорты-v2 или issue в репозитории. В ином случае, опишите баг и ступени для его воспроизведения. -->
<!-- Пример ссылки в Discord-сообщество : https://discord.com/channels/617003227182792704/755125334097133628/ID-сообщения -->

## Тесты
Запустил локалку, вызвал дропбомб 255.  Во время обработки взрыва запустил несколько ракет с малыми взрывами. Все обработалось корректно(насколько это возможно при полном разрушении станции). Понадобится тестмердж, чтобы убедиться в отсутствии критических багов.
<!-- Здесь необходимо описать шаги, которые предпринимались для тестирования изменения. Этот пункт обязателен, без него Pull request будет рассматриваться дольше. -->

dageavtobusnick committed:
experiment: explosions subsystem

dageavtobusnick committed:
experiment: explosions subsystem

Antoonij commented on code/game/atoms.dm:
```diff
@@ -1748,3 +1748,7 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
 	return !(SEND_SIGNAL(src, COMSIG_RIDDEN_DRIVER_MOVE, user, direction) & COMPONENT_DRIVER_BLOCK_MOVE)
+
+///returns how much the object blocks an explosion. Used by subtypes.
+/atom/proc/GetExplosionBlock()
+	CRASH("Unimplemented GetExplosionBlock()")
```
снейк кейс предпочтительнее для названий функций

dageavtobusnick replied:
Эта функция старая, она просто перенесена из объекта в атом.

dageavtobusnick replied:
Хотя, там всего 2 переопределения и 2 вызова. Исправил.

Antoonij commented on code/__HELPERS/data_struct/priority_queue.dm:
```diff
@@ -0,0 +1,52 @@
+    return heap.len == 0
+
+/priority_queue/proc/bubble_up(index)
+    while (index > 1)
+        var/parent = round(index / 2)
+        if (heap[index][1] < heap[parent][1])
+            swap(index, parent)
+            index = parent
+        else
+            break
```
```suggestion
    while(index > 1)
        var/parent = round(index / 2)
        if (heap[index][1] < heap[parent][1])
            swap(index, parent)
            index = parent
        else
            break
```
стиль, также можно сразу и брейкнуть, если куча индекса больше кучи родителя

dageavtobusnick replied:
Исправил

dageavtobusnick committed:
style fix

dageavtobusnick committed:
style fix

dageavtobusnick committed:
fix

dageavtobusnick committed:
fix

dageavtobusnick committed:
infinity cycle fix

dageavtobusnick committed:
infinity cycle fix

dageavtobusnick committed:
Merge branch 'explosion_test' of https://github.com/dageavtobusnick/Paradise into explosion_test

dageavtobusnick committed:
node qdels fix

dageavtobusnick committed:
no_more_infinity_cycle

dageavtobusnick committed:
update

Добавлена корректная более обработка мультиЗ взрывов, теперь взрыв на разны уровнях происходит в рамках одной обработки, а не как последовательность взрывов на разных уровнях.
Так же в прок добавлена возможность содавать направленные взрывы, а так же исключать эпицентр из обработки.
Убрана старая обработка мультиЗ взрывов
